### PR TITLE
[Lang][IR][refactor] Refactor SNode and support coordinate offsets 

### PR DESCRIPTION
**Pull request by yuanming-hu**
Related issue = #938 

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

File: tests/python/test_offset.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    a[1029, 2100, 2200, 2300] = 1
+    assert a[1029, 2100, 2200, 2300] == 1
+
+
+@ti.all_archs
+def test_struct_for():
+    a = ti.var(dt=ti.i32)
+
+    offset = 1024, 2048, 2100, 2200
```
Comment (xumingkuan) [inline tests/python/test_offset.py:18]: Shall we add tests about negative offsets?

Comment (yuanming-hu) [reply tests/python/test_offset.py:18]: Good idea. Added! Waiting for CI to pass.

**Comment by archibate** [comment]
What a cool feature! Could you also update doc for that usage?

**Comment by yuanming-hu** [comment]
Sure, I plan to do that when the coordinate offset feature is at a more mature state, in a separate PR. To prevent myself from forgetting to add a doc, I have added a checklist in #938
